File: src/file-history/backup.ts
import { existsSync } from "node:fs";
import { mkdir, readdir, readFile } from "node:fs/promises";
import { join } from "node:path";
import { getFileHistoryDir } from "../config/paths.ts";

function hashContent(content: string): string {
	const hasher = new Bun.CryptoHasher("md5");
	hasher.update(content);
	return hasher.digest("hex");
}

/**
 * Back up a file's current content before it gets modified.
 * Skips if the file doesn't exist (new file being created).
 * Deduplicates against the latest backup by content hash.
 */
export async function backupFile(filePath: string, projectPath?: string): Promise<void> {
	if (!existsSync(filePath)) return;

	const content = await readFile(filePath, "utf-8");
	const hash = hashContent(content);
	const histDir = getFileHistoryDir(projectPath, filePath);

	// Check latest backup for dedup
	try {
		const files = await readdir(histDir);
		if (files.length > 0) {
			const bakFiles = files.filter((f) => f.endsWith(".bak")).sort();
			if (bakFiles.length > 0) {
				const latest = bakFiles[bakFiles.length - 1] as string;
				const latestContent = await readFile(join(histDir, latest), "utf-8");
				if (hashContent(latestContent) === hash) return;
			}
		}
	} catch {
		// Dir doesn't exist yet — will create below
	}

	await mkdir(histDir, { recursive: true });
	let ts = Date.now();
	// Ensure uniqueness by incrementing if a file with this timestamp already exists
	while (existsSync(join(histDir, `${ts}.bak`))) {
		ts++;
	}
	await Bun.write(join(histDir, `${ts}.bak`), content);
}

File: src/file_history/backup.rs
//! Back up a file's contents before it's modified.

use std::path::Path;

use anyhow::Result;
use sha2::{Digest, Sha256};

use super::meta::FileHistoryMeta;
use crate::config::paths::get_file_history_dir;

fn hash_content(content: &[u8]) -> String {
    let mut hasher = Sha256::new();
    hasher.update(content);
    hex::encode(hasher.finalize())
}

pub fn backup_file(file_path: &Path, project_path: Option<&str>) -> Result<()> {
    if !file_path.exists() {
        return Ok(());
    }
    let content = std::fs::read(file_path)?;
    let hash = hash_content(&content);

    let base_dir = get_file_history_dir(project_path);
    let mut meta = FileHistoryMeta::new(&base_dir);
    let entry = meta.get_or_create(file_path, None)?;
    let uuid_dir = base_dir.join(&entry.uuid);

    if entry.versions > 0 {
        let latest_path = uuid_dir.join(format!("v{}.bak", entry.versions));
        if let Ok(latest) = std::fs::read(&latest_path) {
            if hash_content(&latest) == hash {
                return Ok(());
            }
        }
    }

    std::fs::create_dir_all(&uuid_dir)?;
    let next_version = entry.versions + 1;
    std::fs::write(uuid_dir.join(format!("v{next_version}.bak")), content)?;
    meta.increment_version(&entry.uuid)?;
    Ok(())
}
